File: backend/app/schemas/batch.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class BatchGenerationRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    character_id: str = Field(alias="characterId")
    costume_id: str = Field(alias="costumeId")
    scene_prompts: list[str] = Field(min_length=1, max_length=20, alias="scenePrompts")

    # Shared generation parameters (mirrors GenerationSubmitRequest).
    negative_prompt: str = Field(default="", alias="negativePrompt")
    width: int = Field(ge=64, le=4096, default=1024)
    height: int = Field(ge=64, le=4096, default=1024)
    steps: int = Field(ge=1, le=150, default=28)
    sampler: str = Field(default="DPM++ 2M Karras")
    cfg_scale: float = Field(ge=0.0, le=30.0, alias="cfgScale", default=3.5)
    seed: int | None = None
    lora_weight: float = Field(ge=0.0, le=2.0, alias="loraWeight", default=0.85)
    tags: list[str] = Field(default_factory=list)


class BatchJobItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    task_id: str = Field(alias="taskId")
    scene_prompt: str = Field(alias="scenePrompt")
    status: Literal["pending", "running", "completed", "failed"]
    progress: int = Field(ge=0, le=100)
    message: str | None = None
    error: str | None = None
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")


class BatchGenerationAcceptedResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    batch_id: str = Field(alias="batchId")
    jobs: list[BatchJobItem]
    total: int

File: src/api/batch.ts
const API_BASE_URL = "http://127.0.0.1:8000";

export type BatchJobItem = {
  taskId: string;
  scenePrompt: string;
  status: "pending" | "running" | "completed" | "failed";
  progress: number;
  message?: string | null;
  error?: string | null;
  createdAt: string;
  updatedAt: string;
};

export type BatchGenerationResponse = {
  batchId: string;
  jobs: BatchJobItem[];
  total: number;
};

export type BatchGenerationRequest = {
  characterId: string;
  costumeId: string;
  scenePrompts: string[];
  negativePrompt?: string;
  width: number;
  height: number;
  steps: number;
  sampler: string;
  cfgScale: number;
  seed: number | null;
  loraWeight: number;
  tags?: string[];
};

function isBatchJobItem(value: unknown): value is BatchJobItem {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v["taskId"] === "string" &&
    typeof v["scenePrompt"] === "string" &&
    (v["status"] === "pending" ||
      v["status"] === "running" ||
      v["status"] === "completed" ||
      v["status"] === "failed") &&
    typeof v["progress"] === "number" &&
    typeof v["createdAt"] === "string" &&
    typeof v["updatedAt"] === "string"
  );
}

function isBatchGenerationResponse(value: unknown): value is BatchGenerationResponse {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v["batchId"] === "string" &&
    typeof v["total"] === "number" &&
    Array.isArray(v["jobs"]) &&
    (v["jobs"] as unknown[]).every(isBatchJobItem)
  );
}

export async function submitBatchGeneration(
  request: BatchGenerationRequest,
  signal?: AbortSignal,
): Promise<BatchGenerationResponse> {
  let response: Response;
  try {
    response = await fetch(`${API_BASE_URL}/api/generations/batch`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(request),
      signal,
    });
  } catch {
    throw new Error("BATCH_SERVICE_UNAVAILABLE");
  }

  let payload: unknown;
  try {
    payload = await response.json();
  } catch {
    throw new Error("BATCH_SERVICE_UNAVAILABLE");
  }

  if (!response.ok) {
    const detail =
      typeof payload === "object" &&
      payload !== null &&
      typeof (payload as Record<string, unknown>)["detail"] === "string"
        ? (payload as Record<string, unknown>)["detail"]
        : null;
    throw new Error((detail as string | null) ?? "BATCH_SUBMIT_FAILED");
  }

  if (!isBatchGenerationResponse(payload)) {
    throw new Error("BATCH_INVALID_RESPONSE");
  }

  return payload;
}
